File: cpp/src/Ice/IncomingAsync.cpp
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

#include <Ice/IncomingAsync.h>
#include <Ice/ServantLocator.h>
#include <Ice/Object.h>
#include <Ice/ConnectionI.h>
#include <Ice/LocalException.h>
#include <Ice/Protocol.h>
#include <Ice/Instance.h>
#include <Ice/Properties.h>
#include <Ice/ReplyStatus.h>

#include <mutex>

using namespace std;
using namespace Ice;
using namespace IceInternal;

IceInternal::IncomingAsync::IncomingAsync(Incoming& in) :
    IncomingBase(in),
    _responseHandlerCopy(_responseHandler->shared_from_this())
{
}

shared_ptr<IncomingAsync>
IceInternal::IncomingAsync::create(Incoming& in)
{
    auto async = make_shared<IncomingAsync>(in);
    in.setAsync(async);
    return async;
}

void
IceInternal::IncomingAsync::kill(Incoming& in)
{
    checkResponseSent();
    in._observer.adopt(_observer); // Give back the observer to incoming.
}

void
IceInternal::IncomingAsync::completed()
{
    checkResponseSent();
    IncomingBase::response(true); // User thread
}

void
IceInternal::IncomingAsync::completed(exception_ptr ex)
{
    checkResponseSent();
    IncomingBase::exception(ex, true); // true = amd
}

void
IceInternal::IncomingAsync::checkResponseSent()
{
    if (_responseSent.test_and_set())
    {
        throw ResponseSentException(__FILE__, __LINE__);
    }
}

File: java-compat/src/Ice/src/main/java/IceInternal/IncomingAsync.java
// **********************************************************************
//
// Copyright (c) 2003-present ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package IceInternal;

public class IncomingAsync extends IncomingBase implements Ice.AMDCallback
{
    public
    IncomingAsync(Incoming in)
    {
        super(in);
        in.setAsync(this);
    }

    @Override
    public void
    ice_exception(java.lang.Exception ex)
    {
        if(_interceptorCBs != null)
        {
            try
            {
                for(Ice.DispatchInterceptorAsyncCallback cb : _interceptorCBs)
                {
                    if(!cb.exception(ex))
                    {
                        return;
                    }
                }
            }
            catch(java.lang.RuntimeException exc)
            {
                return;
            }
        }

        checkResponseSent();
        exception(ex, true); // User thread
    }

    final public void
    kill(Incoming in)
    {
        checkResponseSent();

        // Adopt observer and cookie
        in._observer = _observer;
        _observer = null;
        in._cookie = _cookie;
        _cookie = null;
    }

    final protected void
    completed()
    {
        if(_interceptorCBs != null)
        {
            try
            {
                for(Ice.DispatchInterceptorAsyncCallback cb : _interceptorCBs)
                {
                    if(!cb.response())
                    {
                        return;
                    }
                }
            }
            catch(java.lang.RuntimeException exc)
            {
                return;
            }
        }

        checkResponseSent();
        response(true); // User thread
    }

    synchronized private void
    checkResponseSent()
    {
        if(_responseSent)
        {
            throw new Ice.ResponseSentException();
        }
        _responseSent = true;
    }

    private boolean _responseSent = false;
}
